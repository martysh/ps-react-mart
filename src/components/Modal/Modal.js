import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledModal = styled.div`
  position: fixed;
  background: rgba(0, 0, 0, 0.7);
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: none;
  z-index: 2;
  ${props =>
    props.open &&
    `
    display: grid;
    align-items: center;
    justify-items: center;
  `}
  .overlay-inner {
    background: white;
    overflow: auto;
    margin: 48px;
    width: 500px;
    min-height: 30vh;
    max-width: 80vw;
    padding: 2em 3em;
    position: relative;
    box-shadow: 0px 11px 15px -7px rgba(0,0,0,0.2), 0px 24px 38px 3px rgba(0,0,0,0.14), 0px 9px 46px 8px rgba(0,0,0,0.12);
}
  }
  .close {
    position: absolute;
    top: 0.3em;
    right: 0.3em;
    padding: 0.3em;
    font-size: 2rem;
    height: 1em;
    width: 1em;
    text-indent: 10em;
    overflow: hidden;
    background: none;
    color: black;
    border: 0;
    cursor: pointer;
  }
  .close::after{
    position: absolute;
    line-height: 0.5;
    top: 0.2em;
    left: 0.2em;
    text-indent: 0;
    content:'\00D7';
  }
`;

function Modal(props) {
  const { open, toggleModal, children } = props;
  const node = useRef();

  useEffect(() => {
    const handleClick = e => {
      if (!open || node.current.contains(e.target)) return;

      toggleModal();
    };
    document.addEventListener('mousedown', handleClick);
    return () => {
      document.removeEventListener('mousedown', handleClick);
    };
  }, [open]);

  return (
    <StyledModal open={open}>
      <div ref={node} className="overlay-inner">
        <button type="button" className="close" onClick={toggleModal}>Close</button>
        {children}
      </div>
    </StyledModal>
  );
}

Modal.propTypes = {
  open: PropTypes.bool,
  toggleModal: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

Modal.defaultProps = {
  open: false,
};

export default Modal;
