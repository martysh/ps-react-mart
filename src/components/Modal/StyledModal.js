import styled from 'styled-components';

const StyledModal = styled.div`
  position: fixed;
  background: rgba(0, 0, 0, 0.7);
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: none;
  z-index: 2;
  ${props =>
    props.open &&
    `
    display: grid;
    align-items: center;
    justify-items: center;
  `}
  .overlay-inner {
    background: white;
    overflow: auto;
    margin: 48px;
    width: 500px;
    min-height: 30vh;
    max-width: 80vw;
    padding: 2em 3em;
    position: relative;
    box-shadow: 0px 11px 15px -7px rgba(0, 0, 0, 0.2),
      0px 24px 38px 3px rgba(0, 0, 0, 0.14),
      0px 9px 46px 8px rgba(0, 0, 0, 0.12);
  }
  .close {
    position: absolute;
    top: 0.3em;
    right: 0.3em;
    padding: 0.3em;
    font-size: 2rem;
    height: 1em;
    width: 1em;
    text-indent: 10em;
    overflow: hidden;
    background: none;
    color: black;
    border: 0;
    cursor: pointer;
  }
  .close:after {
    position: absolute;
    line-height: 0.5;
    top: 0.2em;
    left: 0.2em;
    text-indent: 0;
    content: '\00D7';
  }
`;

export default StyledModal;
