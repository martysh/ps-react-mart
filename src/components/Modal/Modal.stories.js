import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import Modal from './Modal';
import Button from '../Button';

storiesOf('Modal', module).add('basic modal', () => {
  function ModalExample() {
    const [showModal, toggleModal] = useState(false);
    const onToggleModal = () => {
      toggleModal(isOpen => !isOpen);
    };
    return (
      <>
        <Button onClick={onToggleModal}>Toggle my Modal</Button>
        <Modal open={showModal} toggleModal={onToggleModal}>
          well hello from the modal!
        </Modal>
      </>
    );
  }
  return <ModalExample />;
});
