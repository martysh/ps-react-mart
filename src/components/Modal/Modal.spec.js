import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import Modal from './Modal';
import 'jest-styled-components';

Enzyme.configure({ adapter: new Adapter() });

describe('the Modal component', () => {
  fit('toggleClick works', () => {
    const onCloseClick = jest.fn();

    const wrapper = shallow(
      <Modal open toggleModal={onCloseClick}>
        <div className="content">Hello World</div>
      </Modal>
    );
    // expect(wrapper).toHaveStyleRule('display', 'none');
    expect(wrapper.find('.content')).toHaveLength(1);
    expect(wrapper.find('.content').text()).toBe('Hello World');

    wrapper.find('.close').simulate('click');
    expect(onCloseClick).toHaveBeenCalled();
  });

  it('should match the last snapshot', () => {
    const wrapper = shallow(
      <Modal open toggleModal={() => {}}>
        nothing
      </Modal>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('it closes on toggle', () => {
    const tree = renderer.create(
      <Modal key={1} open toggleModal={() => {}}>
        {' '}
        nothing
      </Modal>
    );
    const jsonTree = tree.toJSON();
    expect(jsonTree).toHaveStyleRule('display', 'grid');
    tree.update(
      <Modal key={2} open={false} toggleModal={() => {}}>
        nothing
      </Modal>
    );
    expect(tree.toJSON()).toHaveStyleRule('display', 'none');
  });

  it('it works with react test renderer', () => {
    const tree = renderer
      .create(
        <Modal open toggleModal={() => {}}>
          nothing
        </Modal>
      )
      .toJSON();
    expect(tree).toHaveStyleRule('display', 'grid');
    expect(tree).toMatchSnapshot();
  });
});
