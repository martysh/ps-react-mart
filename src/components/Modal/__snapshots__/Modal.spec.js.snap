// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Modal component it works with react test renderer 1`] = `
.c0 {
  position: fixed;
  background: rgba(0,0,0,0.7);
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: none;
  z-index: 2;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
}

.c0 .overlay-inner {
  background: white;
  overflow: auto;
  margin: 48px;
  width: 500px;
  min-height: 30vh;
  max-width: 80vw;
  padding: 2em 3em;
  position: relative;
  box-shadow: 0px 11px 15px -7px rgba(0,0,0,0.2), 0px 24px 38px 3px rgba(0,0,0,0.14), 0px 9px 46px 8px rgba(0,0,0,0.12);
}

.c0 .close {
  position: absolute;
  top: 0.3em;
  right: 0.3em;
  padding: 0.3em;
  font-size: 2rem;
  height: 1em;
  width: 1em;
  text-indent: 10em;
  overflow: hidden;
  background: none;
  color: black;
  border: 0;
  cursor: pointer;
}

.c0 .close:after {
  position: absolute;
  line-height: 0.5;
  top: 0.2em;
  left: 0.2em;
  text-indent: 0;
  content: '\\0D7';
}

<div
  className="c0"
  open={true}
>
  <div
    className="overlay-inner"
  >
    <button
      className="close"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
    nothing
  </div>
</div>
`;

exports[`the Modal component should match the last snapshot 1`] = `
<styled.div
  open={true}
>
  <div
    className="overlay-inner"
  >
    <button
      className="close"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
    nothing
  </div>
</styled.div>
`;
